stock_resupply_order/models/stock_resupply_order.py

```diff
@@ -0,0 +1,191 @@
+            # service products have no tracking/lot and cannot be run in procurement
+            if line.product_id.product_tmpl_id.type == "service":
+                continue
+
+            available_quantity = self._get_available_quantity_for_product(
+                quant_groups, line
+            )
+
+            if available_quantity < line.quantity:
+                procurements.append(
+                    self.env["procurement.group"].Procurement(
+                        product_id=line.product_id,
+                        product_qty=line.quantity - available_quantity,
+                        product_uom=line.product_id.product_tmpl_id.uom_id,
+                        location_id=self.location_id,
+                        name=f"Resupply Order to {self.location_id.name}",
+                        origin=f"Resupply Order to {self.location_id.name}",
+                        company_id=self.company_id,
+                        values=self._get_procurement_values(),
+                    )
+                )
+
+        if procurements:
+            self.env["procurement.group"].run(procurements)
+
+        self.state = "done"
+
+        return self.procurement_group_id
+
+    @api.model
+    def _get_available_quantity_for_product(
+        self, quant_groups, stock_resupply_order_line
+    ):
+        try:
+            # I could not find a way to merge stock_resupply_order_lines
+            # with the quant_groups query, so it is retrieved with a product
+            # id search here. Not ideal, but it works.
+            group = next(
+                group
+                for group in quant_groups
+                if group["product_id"][0] == stock_resupply_order_line.product_id.id
+            )
+
+            return group["quantity"] - group["reserved_quantity"]
+        except StopIteration:
+            return 0
+
+    def _get_procurement_values(self):
+        """
+        Values to pass to the procurement once the order is run.
+        """
+
+        self.ensure_one()
+
+        return {
+            "group_id": self.procurement_group_id,
+        }
+
+    def _get_existing_quants(self):
+        """
+        Get stock quants at the targeted location. Override if you need to
+        apply specific constraints.
+        """
+
+        self.ensure_one()
+
+        return (
+            self.env["stock.quant"]
+            .sudo()
+            .read_group(
+                domain=[
+                    ("location_id", "=", self.location_id.id),
+                    (
+                        "product_id",
+                        "in",
+                        self.stock_resupply_order_lines.product_id.ids,
+                    ),
+                ],
+                # Cant aggregate available_quantity here.
+                fields=["product_id", "quantity:sum", "reserved_quantity:sum"],
+                groupby=["product_id"],
+                orderby="product_id",
+                lazy=False,
+            )
+        )
```

Comment:
I suppose other part of the source code would be easier to read if this return a dict (using a default dict) some thing likes:

```python
available_quantities: dict[int, int] = defaultdict(int)
for group in ...read_group():
   product_id = group...
   available_quantity = physic - reserved
   available_quantities[product_id] = available_quantity
``